File: src/main/java/xyz/baddeveloper/lwsl/server/SocketServer.java
package xyz.baddeveloper.lwsl.server;

import dev.throwouterror.eventbus.SimpleEventManager;
import xyz.baddeveloper.lwsl.server.events.ServerConnectEvent;
import xyz.baddeveloper.lwsl.server.events.ServerDisconnectEvent;
import xyz.baddeveloper.lwsl.server.events.ServerReadyEvent;

import javax.net.ssl.SSLContext;
import java.io.IOException;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Executors;

public class SocketServer {

    private static final int DEFAULT_PORT = 8080;

    private int port;
    private int maxconnections;
    private int timeout;

    final SimpleEventManager eventManager = new SimpleEventManager();

    private List<SocketHandler> clients = new ArrayList<>();

    private boolean running = false;
    private Controller controller;
    private ServerSocket serverSocket;
    private final SSLContext sslContext;

    public SocketServer() {
        this(DEFAULT_PORT);
    }

    public SocketServer(SSLContext sslContext) {
        this(DEFAULT_PORT, sslContext);
    }

    public SocketServer(int port) {
        this(port, null);
    }

    public SocketServer(int port, SSLContext sslContext) {
        this.port = port;
        this.maxconnections = 1000;
        this.timeout = 0;
        this.sslContext = sslContext;
    }

    public void start() {
        try {
            if (sslContext != null) serverSocket = sslContext.getServerSocketFactory().createServerSocket(port);
            else serverSocket = new ServerSocket(port);

            serverSocket.setSoTimeout(timeout);
        } catch (IOException e) {
            e.printStackTrace();
        }
        running = serverSocket != null;
        listen();
    }

    private void listen() {
        if (!running)
            return;
        controller = new Controller(this);

        Executors.newSingleThreadExecutor().execute(() -> {
            SocketHandler socketHandler = null;
            while (running) try {
                Socket socket = serverSocket.accept();
                socketHandler = new SocketHandler(this, socket);
                if (maxconnections != 0 && clients.size() >= maxconnections) {
                    eventManager.fireEvent(new ServerDisconnectEvent(this, socketHandler));
                    socket.close();
                    return;
                }

                clients.add(socketHandler);
                eventManager.fireEvent(new ServerConnectEvent(this, socketHandler));
                socketHandler.handle();
            } catch (IOException ignored) {
                eventManager.fireEvent(new ServerDisconnectEvent(this, socketHandler));
            }
        });
        eventManager.fireEvent(new ServerReadyEvent(this));
    }

    public void stop() {
        if (serverSocket != null)
            try {
                serverSocket.close();
            } catch (IOException ignored) {
            }
    }

    public SocketServer addEventListener(Object eventHandlerClass) {
        eventManager.addEventListener(eventHandlerClass);
        return this;
    }

    public SocketServer setPort(int port) {
        this.port = port;
        return this;
    }

    public SocketServer setMaxConnections(int maxConnections) {
        this.maxconnections = maxConnections;
        return this;
    }

    public SocketServer setTimeout(int timeout) {
        this.timeout = timeout;
        return this;
    }

    public boolean isRunning() {
        return running;
    }

    public int getPort() {
        return port;
    }

    public int getMaxconnections() {
        return maxconnections;
    }

    public int getTimeout() {
        return timeout;
    }

    public ServerSocket getServerSocket() {
        return serverSocket;
    }

    public List<SocketHandler> getClients() {
        return clients;
    }

    public Controller getController() {
        return controller;
    }
}

File: src/main/java/com/firenodes/lwsl/server/SocketServer.kt
package com.firenodes.lwsl.server

import dev.throwouterror.eventbus.TypedEventBus
import io.netty.handler.logging.LogLevel
import io.netty.handler.timeout.ReadTimeoutHandler
import reactor.netty.DisposableServer
import reactor.netty.tcp.SslProvider
import reactor.netty.tcp.TcpServer
import com.firenodes.lwsl.Events
import com.firenodes.lwsl.exceptions.ConnectException
import com.firenodes.lwsl.packet.ClientIdentificationPacket
import com.firenodes.lwsl.server.events.ServerEvent
import java.io.IOException
import java.lang.Exception
import java.util.concurrent.TimeUnit

class SocketServer @JvmOverloads constructor(var port: Int = DEFAULT_PORT, private val sslProvider: SslProvider? = null) {
    var maxconnections = DEFAULT_MAX_CONNECTIONS
        private set
    var timeout = 0
        private set
    val eventManager = TypedEventBus<ServerEvent>()
    val clients: MutableMap<String, SocketHandler?> = mutableMapOf()
    var isRunning = false
        private set
    var controller: Controller? = null
        private set
    var serverSocket: TcpServer = TcpServer.create()
        private set
    lateinit var disposableServer: DisposableServer
        private set
    private var clientCreationHandler: ClientCreationHandler = object : ClientCreationHandler {
        override fun createClient(socketHandler: SocketHandler) {
            socketHandler.sendPacket(ClientIdentificationPacket(socketHandler.id))
            clients[socketHandler.id] = socketHandler
        }
    }

    constructor(sslContext: SslProvider?) : this(DEFAULT_PORT, sslContext) {}

    fun start(): SocketServer {
        isRunning = true
        listen()
        return this
    }

    private fun listen() {
        if (!isRunning) return
        controller = Controller(this)
        try {
            serverSocket.doOnConnection { socket ->
                var socketHandler: SocketHandler? = null
                try {
                    socket.addHandler(ReadTimeoutHandler(timeout.toLong(), TimeUnit.SECONDS))
                    socketHandler = SocketHandler(this, socket)
                    if (maxconnections != 0 && clients.size >= maxconnections) {
                        eventManager.fireEvent(ServerEvent(Events.DISCONNECT.toString(), this, socketHandler))
                        socket.dispose()
                    }
                    clientCreationHandler.createClient(socketHandler)
                    eventManager.fireEvent(ServerEvent(Events.CONNECT.toString(), this, socketHandler))
                    socketHandler.handle()
                } catch (ignored: IOException) {
                    eventManager.fireEvent(ServerEvent(Events.DISCONNECT.toString(), this, socketHandler))
                    if (socketHandler != null) {
                        clients.remove(socketHandler.id)
                    }
                }
            }.doOnBound {
                eventManager.fireEvent(ServerEvent(Events.READY.toString(), this))
            }.host("0.0.0.0").port(port)

            if (sslProvider != null) serverSocket.secure(sslProvider)
            disposableServer = serverSocket.bindNow()
        } catch (e: Exception) {
            throw ConnectException(e)
        }
    }

    fun enableDebugLogging(): SocketServer {
        serverSocket.wiretap("LWSL", LogLevel.INFO)
        return this
    }

    fun handleClientCreation(handler: ClientCreationHandler): SocketServer {
        clientCreationHandler = handler
        return this
    }

    fun stop() {
        disposableServer.disposeNow()
    }

    fun createEventBus(): TypedEventBus<ServerEvent> {
        return eventManager
    }

    fun setPort(port: Int): SocketServer {
        this.port = port
        return this
    }

    fun setMaxConnections(maxConnections: Int): SocketServer {
        maxconnections = maxConnections
        return this
    }

    fun setTimeout(timeout: Int): SocketServer {
        this.timeout = timeout
        return this
    }

    companion object {
        const val DEFAULT_PORT = 8080
        const val DEFAULT_MAX_CONNECTIONS = 1000
    }
}
